File: src/mcp_server_check/tools/external_payrolls.py
"""External payroll tools for the Check API."""

from __future__ import annotations

from mcp.server.fastmcp import FastMCP

from mcp_server_check.helpers import (
    Ctx,
    check_api_delete,
    check_api_get,
    check_api_list,
    check_api_patch,
    check_api_post,
)


async def list_external_payrolls(
    ctx: Ctx,
    limit: int | None = None,
    ids: list[str] | None = None,
    cursor: str | None = None,
) -> dict:
    """List external payrolls across all companies.

    Args:
        limit: Maximum number of results to return (default 10, max 100).
        ids: Filter to specific external payroll IDs.
        cursor: Pagination cursor from a previous response.
    """
    params: dict = {}
    if limit is not None:
        params["limit"] = limit
    if ids:
        params["ids"] = ",".join(ids)
    if cursor:
        params["cursor"] = cursor
    return await check_api_list(ctx, "/external_payrolls", params=params or None)


async def get_external_payroll(ctx: Ctx, payroll_id: str) -> dict:
    """Get details for a specific external payroll.

    Args:
        payroll_id: The Check external payroll ID.
    """
    return await check_api_get(ctx, f"/external_payrolls/{payroll_id}")


async def create_external_payroll(
    ctx: Ctx,
    company: str,
    period_start: str,
    period_end: str,
    payday: str,
    data: dict | None = None,
) -> dict:
    """Create a new external payroll.

    Args:
        company: The Check company ID.
        period_start: Pay period start date (YYYY-MM-DD).
        period_end: Pay period end date (YYYY-MM-DD).
        payday: Payday date (YYYY-MM-DD).
        data: Additional external payroll fields.
    """
    body: dict = {
        "company": company,
        "period_start": period_start,
        "period_end": period_end,
        "payday": payday,
    }
    if data:
        body.update(data)
    return await check_api_post(ctx, "/external_payrolls", data=body)


async def update_external_payroll(ctx: Ctx, payroll_id: str, data: dict) -> dict:
    """Update an existing external payroll.

    Args:
        payroll_id: The Check external payroll ID.
        data: Fields to update.
    """
    return await check_api_patch(ctx, f"/external_payrolls/{payroll_id}", data=data)


async def delete_external_payroll(ctx: Ctx, payroll_id: str) -> dict:
    """Delete an external payroll.

    Args:
        payroll_id: The Check external payroll ID.
    """
    return await check_api_delete(ctx, f"/external_payrolls/{payroll_id}")


async def approve_external_payroll(ctx: Ctx, payroll_id: str) -> dict:
    """Approve an external payroll.

    Args:
        payroll_id: The Check external payroll ID.
    """
    return await check_api_post(ctx, f"/external_payrolls/{payroll_id}/approve")


async def reopen_external_payroll(ctx: Ctx, payroll_id: str) -> dict:
    """Reopen an external payroll.

    Args:
        payroll_id: The Check external payroll ID.
    """
    return await check_api_post(ctx, f"/external_payrolls/{payroll_id}/reopen")


async def preview_external_payroll(ctx: Ctx, external_payroll_id: str) -> dict:
    """Preview an external payroll.

    Args:
        external_payroll_id: The Check external payroll ID.
    """
    return await check_api_get(ctx, f"/external_payrolls/{external_payroll_id}/preview")


def register(mcp: FastMCP) -> None:
    mcp.add_tool(list_external_payrolls)
    mcp.add_tool(get_external_payroll)
    mcp.add_tool(create_external_payroll)
    mcp.add_tool(update_external_payroll)
    mcp.add_tool(delete_external_payroll)
    mcp.add_tool(approve_external_payroll)
    mcp.add_tool(reopen_external_payroll)
    mcp.add_tool(preview_external_payroll)

File: ts/src/tools/external-payrolls.ts
/** External payroll tools for the Check API. */

import { z } from "zod";
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import {
  checkApiGet,
  checkApiPost,
  checkApiPatch,
  checkApiDelete,
  checkApiList,
  type CheckApiOptions,
} from "../helpers.js";
import type { ToolFilter } from "../tool-filter.js";
import { tool } from "./_register.js";



export function register(
  server: McpServer,
  api: CheckApiOptions,
  filter: ToolFilter,
  toolsetName: string,
) {
  tool(
    server,
    filter,
    toolsetName,
    "list_external_payrolls",
    "List external payrolls, optionally filtered by company.",
    {
      company: z.string().optional().describe('Filter to external payrolls belonging to this Check company ID (e.g. "com_xxxxx").'),
      limit: z.number().optional().describe("Maximum number of results to return (default 10, max 100)."),
      ids: z.array(z.string()).optional().describe("Filter to specific external payroll IDs."),
      cursor: z.string().optional().describe("Pagination cursor from a previous response."),
    },
    async ({ company, limit, ids, cursor }) => {
      const params: Record<string, unknown> = {};
      if (company !== undefined) params.company = company;
      if (limit !== undefined) params.limit = limit;
      if (ids !== undefined) params.ids = ids.join(",");
      if (cursor !== undefined) params.cursor = cursor;
      const result = await checkApiList(api, "/external_payrolls", params);
      return { content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }] };
    },
  );

  tool(
    server,
    filter,
    toolsetName,
    "get_external_payroll",
    "Get details for a specific external payroll.",
    {
      payroll_id: z.string().describe("The Check external payroll ID."),
    },
    async ({ payroll_id }) => {
      const result = await checkApiGet(api, `/external_payrolls/${payroll_id}`);
      return { content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }] };
    },
  );

  tool(
    server,
    filter,
    toolsetName,
    "create_external_payroll",
    "Create a new external payroll.",
    {
      company: z.string().describe("The Check company ID."),
      period_start: z.string().describe("Pay period start date (YYYY-MM-DD)."),
      period_end: z.string().describe("Pay period end date (YYYY-MM-DD)."),
      payday: z.string().describe("Payday date (YYYY-MM-DD)."),
      pay_frequency: z.string().optional().describe("Frequency at which the external payroll was paid."),
      items: z.array(z.record(z.unknown())).optional().describe('List of external payroll item objects. Each may include "employee", "earnings" (list), "reimbursements" (list), "taxes" (list), "benefits" (list), "post_tax_deductions" (list).'),
      contractor_payments: z.array(z.record(z.unknown())).optional().describe('List of contractor payment objects. Each may include "contractor", "amount", "reimbursement_amount".'),
    },
    async ({ company, period_start, period_end, payday, pay_frequency, items, contractor_payments }) => {
      const body: Record<string, unknown> = { company, period_start, period_end, payday };
      if (pay_frequency !== undefined) body.pay_frequency = pay_frequency;
      if (items !== undefined) body.items = items;
      if (contractor_payments !== undefined) body.contractor_payments = contractor_payments;
      const result = await checkApiPost(api, "/external_payrolls", body);
      return { content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }] };
    },
  );

  tool(
    server,
    filter,
    toolsetName,
    "update_external_payroll",
    "Update an existing external payroll.",
    {
      payroll_id: z.string().describe("The Check external payroll ID."),
      period_start: z.string().optional().describe("Pay period start date (YYYY-MM-DD)."),
      period_end: z.string().optional().describe("Pay period end date (YYYY-MM-DD)."),
      payday: z.string().optional().describe("Payday date (YYYY-MM-DD)."),
      pay_frequency: z.string().optional().describe("Frequency at which the external payroll was paid."),
      items: z.array(z.record(z.unknown())).optional().describe("List of external payroll item objects (see create_external_payroll)."),
      contractor_payments: z.array(z.record(z.unknown())).optional().describe("List of contractor payment objects (see create_external_payroll)."),
    },
    async ({ payroll_id, period_start, period_end, payday, pay_frequency, items, contractor_payments }) => {
      const body: Record<string, unknown> = {};
      if (period_start !== undefined) body.period_start = period_start;
      if (period_end !== undefined) body.period_end = period_end;
      if (payday !== undefined) body.payday = payday;
      if (pay_frequency !== undefined) body.pay_frequency = pay_frequency;
      if (items !== undefined) body.items = items;
      if (contractor_payments !== undefined) body.contractor_payments = contractor_payments;
      const result = await checkApiPatch(api, `/external_payrolls/${payroll_id}`, body);
      return { content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }] };
    },
  );

  tool(
    server,
    filter,
    toolsetName,
    "delete_external_payroll",
    "Delete an external payroll.",
    {
      payroll_id: z.string().describe("The Check external payroll ID."),
    },
    async ({ payroll_id }) => {
      const result = await checkApiDelete(api, `/external_payrolls/${payroll_id}`);
      return { content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }] };
    },
  );

  tool(
    server,
    filter,
    toolsetName,
    "approve_external_payroll",
    "Approve an external payroll.",
    {
      payroll_id: z.string().describe("The Check external payroll ID."),
    },
    async ({ payroll_id }) => {
      const result = await checkApiPost(api, `/external_payrolls/${payroll_id}/approve`);
      return { content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }] };
    },
  );

  tool(
    server,
    filter,
    toolsetName,
    "reopen_external_payroll",
    "Reopen an external payroll.",
    {
      payroll_id: z.string().describe("The Check external payroll ID."),
    },
    async ({ payroll_id }) => {
      const result = await checkApiPost(api, `/external_payrolls/${payroll_id}/reopen`);
      return { content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }] };
    },
  );

  tool(
    server,
    filter,
    toolsetName,
    "preview_external_payroll",
    "Preview an external payroll.",
    {
      external_payroll_id: z.string().describe("The Check external payroll ID."),
    },
    async ({ external_payroll_id }) => {
      const result = await checkApiGet(api, `/external_payrolls/${external_payroll_id}/preview`);
      return { content: [{ type: "text" as const, text: JSON.stringify(result, null, 2) }] };
    },
  );
}
